django/thunderstore/community/models/package_listing.py

```diff
@@ -19,6 +20,17 @@ def active(self):
             ~Q(package__versions__is_active=True)
         )
 
+    def annotate_canonical_name(self):
+        return self.annotate(
+            canonical_name=Concat(
+                F("package__owner__name"),
+                Value("-"),
+                F("package__name"),
+                Value("-"),
+                F("package__versions__version_number"),
```

Comment:
Does this really pick the right version? What happens when there are multiple versions with a package?

Reply:
This doesn't pick anything, it just annotates all the packages in the QuerySet. If you need to pick something specific, use filters/excludes. In this use case we actually want all the versions, since we want to validate the version too.

Reply:
I had assumed the annotation is on a per-row basis, and this'd end up picking the first/last version to annotate to the row, but I now saw it's being used with `.only()` and perhaps Django queries it in a fashion where it generates an unique row for each annotation, rather than just one per package?

Reply:
So after rechecking the queries, the `.only()` didn't do what I expected it to do without any parameters, so I changed it to `.only("id")`. The difference being that the former selects all the fields from `community_packagelisting`, while the latter only picks the id and the annotated field. IIRC I originally added this to avoid querying the joined tables and including their fields in the results since we didn't need them, but seems they're not picked anyway.

Maybe the query resulting from `community.package_listings.active().annotate_canonical_name().only("id")` helps to clarify things:
```python
SELECT
    "community_packagelisting"."id",
    CONCAT(
        "repository_team"."name", CONCAT(
            \'-\', CONCAT(
                "repository_package"."name", CONCAT(
                    \'-\', "repository_packageversion"."version_number"
                )
            )
        )
    ) AS "canonical_name"
FROM "community_packagelisting"
    INNER JOIN "repository_package" ON ("community_packagelisting"."package_id" = "repository_package"."id")
    INNER JOIN "repository_team" ON ("repository_package"."owner_id" = "repository_team"."id")
    LEFT OUTER JOIN "repository_packageversion" ON ("repository_package"."id" = "repository_packageversion"."package_id")
WHERE (
    "community_packagelisting"."community_id" = 2
    AND NOT (NOT "repository_package"."is_active")
    AND NOT (
        NOT (
            "community_packagelisting"."package_id" IN (
                SELECT U2."package_id" FROM "repository_packageversion" U2 WHERE U2."is_active"
            )
        )
    )
)
```